File: gpc_agents/due_diligence.py
"""
Gallagher Property Company - Due Diligence Agent
"""

from functools import partial
from typing import Any, Dict, List, Optional

from agents import Agent, WebSearchTool
from agents import function_tool as base_function_tool
from pydantic import BaseModel, Field

from config.settings import settings
from prompts.agent_prompts import DUE_DILIGENCE_PROMPT
from tools.database import db

function_tool = partial(base_function_tool, strict_mode=False)


class CreateDdDealInput(BaseModel):
    """Input for creating a due diligence deal"""

    project_id: Optional[str] = None
    name: str
    status: str = "open"
    key_dates: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class IngestDdDocumentInput(BaseModel):
    """Input for ingesting a due diligence document"""

    dd_deal_id: str
    document_type: str
    title: Optional[str] = None
    storage_ref: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class GenerateDdChecklistInput(BaseModel):
    """Input for generating a due diligence checklist"""

    dd_deal_id: str
    property_type: str
    phase: str


class FlagDdRedFlagsInput(BaseModel):
    """Input for recording red flags"""

    dd_deal_id: str
    findings: List[Dict[str, Any]] = Field(default_factory=list)


class SaveDdSummaryInput(BaseModel):
    """Input for saving DD summary"""

    project_id: Optional[str] = None
    dd_deal_id: Optional[str] = None
    summary: str
    recommendation: str
    confidence: str = "medium"
    supporting_data: Dict[str, Any] = Field(default_factory=dict)


_CHECKLIST_TEMPLATES = {
    "acquisition": [
        "Review title report and vesting deeds",
        "Order Phase I environmental report",
        "Collect rent roll and operating statements",
        "Verify zoning compliance and permitted uses",
    ],
    "development": [
        "Confirm utility availability and capacity",
        "Collect survey, ALTA, and boundary details",
        "Review entitlements timeline and fee schedule",
        "Validate construction budget and GMP",
    ],
    "operations": [
        "Inspect physical condition and deferred maintenance",
        "Confirm insurance coverage and claims history",
        "Review vendor contracts and service agreements",
        "Assess market comps and leasing velocity",
    ],
}


def _build_checklist(property_type: str, phase: str) -> List[Dict[str, Any]]:
    base_items = _CHECKLIST_TEMPLATES.get(phase.lower(), _CHECKLIST_TEMPLATES["acquisition"])
    return [
        {
            "phase": phase,
            "name": item,
            "status": "pending",
            "metadata": {"property_type": property_type},
        }
        for item in base_items
    ]


@function_tool
async def create_dd_deal(input_data: CreateDdDealInput) -> Dict[str, Any]:
    """
    Create a new due diligence deal record

    Args:
        input_data: Deal metadata

    Returns:
        Created deal record
    """
    record = {
        "project_id": input_data.project_id,
        "name": input_data.name,
        "status": input_data.status,
        "key_dates": input_data.key_dates,
        "metadata": input_data.metadata,
    }
    deal = await db.create_dd_deal(record)
    return {"deal": deal}


@function_tool
async def ingest_dd_document(input_data: IngestDdDocumentInput) -> Dict[str, Any]:
    """
    Ingest a due diligence document

    Args:
        input_data: Document metadata

    Returns:
        Created document record
    """
    record = {
        "dd_deal_id": input_data.dd_deal_id,
        "document_type": input_data.document_type,
        "title": input_data.title,
        "storage_ref": input_data.storage_ref,
        "metadata": input_data.metadata,
    }
    doc = await db.add_dd_document(record)
    return {"document": doc}


@function_tool
async def generate_dd_checklist(input_data: GenerateDdChecklistInput) -> Dict[str, Any]:
    """
    Generate a due diligence checklist

    Args:
        input_data: DD deal, property type, and phase

    Returns:
        Checklist items inserted
    """
    items = _build_checklist(input_data.property_type, input_data.phase)
    stored = await db.add_dd_checklist_items(input_data.dd_deal_id, items)
    return {"items": stored, "count": len(stored)}


@function_tool
async def flag_dd_red_flags(input_data: FlagDdRedFlagsInput) -> Dict[str, Any]:
    """
    Record red flags for a DD deal

    Args:
        input_data: Findings list

    Returns:
        Red flags inserted
    """
    normalized = []
    for finding in input_data.findings:
        description = finding.get("description") or finding.get("details") or finding.get("title")
        normalized.append(
            {
                "dd_deal_id": input_data.dd_deal_id,
                "severity": finding.get("severity", "medium"),
                "description": description,
                "category": finding.get("category"),
                "status": finding.get("status", "open"),
                "metadata": {
                    "title": finding.get("title"),
                    "details": finding.get("details"),
                    "source": finding.get("source"),
                },
            }
        )
    flags = await db.add_dd_red_flags(input_data.dd_deal_id, normalized)
    return {"red_flags": flags, "count": len(flags)}


@function_tool
async def save_dd_summary(input_data: SaveDdSummaryInput) -> Dict[str, Any]:
    """Save DD summary to agent_outputs"""
    output = await db.save_agent_output(
        {
            "project_id": input_data.project_id,
            "agent_name": "due_diligence",
            "task_type": "dd_summary",
            "input_data": {"dd_deal_id": input_data.dd_deal_id},
            "output_data": {
                "summary": input_data.summary,
                "recommendation": input_data.recommendation,
                "supporting_data": input_data.supporting_data,
            },
            "confidence": input_data.confidence,
        }
    )
    return {"saved": True, "output": output}


# Agent definition

due_diligence_agent = Agent(
    name="Due Diligence",
    instructions=DUE_DILIGENCE_PROMPT,
    tools=[
        create_dd_deal,
        ingest_dd_document,
        generate_dd_checklist,
        flag_dd_red_flags,
        save_dd_summary,
        WebSearchTool(),
    ],
    model=settings.openai.flagship_model,
)

File: packages/openai/src/agents/dueDiligence.ts
import { Agent } from '@openai/agents';
import { AGENT_MODEL_IDS } from '@entitlement-os/shared';

export const dueDiligenceAgent = new Agent({
  name: 'Due Diligence',
  model: AGENT_MODEL_IDS.dueDiligence,
  handoffDescription:
    'Tracks diligence items, catalogs documents, generates checklists, flags red flags, summarizes readiness for investment decisions, and queries document extractions',
  instructions: `You are the Due Diligence Coordinator for Gallagher Property Company. Your role is to track diligence items, capture documents, flag red flags, and summarize readiness for investment decisions.

## CORE RESPONSIBILITIES
1. Create and manage due diligence deals
2. Ingest and catalog diligence documents
3. Generate phase-specific checklists
4. Track red flags with severity and status
5. Summarize findings and next steps

## DOCUMENT INTELLIGENCE PROTOCOL
You have direct access to the document extraction pipeline. Use it aggressively:

1. **Start every DD assessment with get_document_extraction_summary** - this tells you what documents have been uploaded and extracted, which doc types are present, average confidence levels, and how many extractions are still unreviewed.
2. **Query specific extractions with query_document_extractions**:
- \`doc_type: "phase_i_esa"\` - check for RECs, de minimis conditions, Phase II recommendations
- \`doc_type: "title_commitment"\` - review exceptions, easements, liens, encumbrances
- \`doc_type: "survey"\` - flood zone, acreage, encroachments, setbacks
- \`doc_type: "psa"\` - purchase price, DD period, contingencies, closing date
- \`doc_type: "appraisal"\` - appraised value vs purchase price gap
- \`doc_type: "financing_commitment"\` - lender terms, conditions, expiry date
- \`doc_type: "zoning_letter"\` - current zoning, permitted/conditional uses, variances
- \`doc_type: "lease"\` - tenant terms, escalation structure, renewal options
3. **Use compare_document_vs_deal_terms** to catch mismatches between what documents say and what's modeled in the deal.
4. **Check for unreviewed extractions** - flag any extraction with \`reviewed: false\` and \`confidence < 0.85\` as needing human verification before relying on it for investment decisions.

### DD Readiness Assessment via Documents
When assessing deal readiness, check for the presence and quality of:

| Required Document | Doc Type Key | Critical Fields |
|---|---|---|
| Purchase Agreement | \`psa\` | purchase_price, dd_period, closing_date, contingencies |
| Title Commitment | \`title_commitment\` | exceptions, liens, easements |
| Phase I ESA | \`phase_i_esa\` | recs, recommended_phase_ii |
| Survey/ALTA | \`survey\` | flood_zone, acreage, encroachments |
| Appraisal | \`appraisal\` | appraised_value, cap_rate |
| Financing Commitment | \`financing_commitment\` | loan_amount, rate, conditions |
| Zoning Verification | \`zoning_letter\` | current_zoning, permitted_uses |

If any of these are missing, flag them as DD gaps.

## QUALITY BAR (CONSULTING GRADE)
- Separate facts from assumptions; do not infer without evidence
- Clearly enumerate critical gaps blocking investment decisions
- Provide a prioritized next-step list with owners/timelines if available
- Include confidence level and rationale

## DD PHASES & CHECKLIST TEMPLATES

### Acquisition Phase
- Review title report and vesting deeds
- Order Phase I environmental report
- Collect rent roll and operating statements
- Verify zoning compliance and permitted uses

### Development Phase
- Confirm utility availability and capacity
- Collect survey, ALTA, and boundary details
- Review entitlements timeline and fee schedule
- Validate construction budget and GMP

### Operations Phase
- Inspect physical condition and deferred maintenance
- Confirm insurance coverage and claims history
- Review vendor contracts and service agreements
- Assess market comps and leasing velocity

## RED FLAG SEVERITY LEVELS
| Severity | Description | Action |
|----------|-------------|--------|
| Critical | Deal-breaker potential | Halt process, escalate immediately |
| High | Material financial impact | Investigate before proceeding |
| Medium | Notable concern | Monitor and mitigate |
| Low | Minor observation | Document and track |

## OUTPUT FORMAT

### DD Status Summary
**Deal:** [Name]
**Phase:** [Acquisition/Development/Operations]
**Status:** [Open/In Progress/Complete]

**Document Coverage:**
| Document | Status | Confidence | Key Findings |
|----------|--------|------------|--------------|
| [Doc type] | [Extracted/Missing/Needs Review] | [X%] | [Summary] |

**Checklist Progress:** X/Y items complete

**Red Flags:**
| # | Severity | Description | Status |
|---|----------|-------------|--------|
| 1 | [Critical/High/Medium/Low] | [Description] | [Open/Resolved] |

**Critical Gaps:**
- [Gap blocking investment decision]

**Recommendation:** [Proceed/Conditional/Pass]
**Confidence:** [High/Medium/Low]
**Next Steps:**
1. [Action] - [Owner] - [Due Date]`,
  tools: [],
  handoffs: [],
});
